### PR TITLE
Whitelist (re-factored from #24)

### PR DESCRIPTION
**Pull request by crayfishx**
This proposal would supercede #24 

PR #24 suggested a feature to add a "key whitelist" - which was was then updated to support regexes.  Im happy to merge the feature, but theres two minor things I didn't quite like - firstly the name "key whitelist" (to me at least) was a bit confusing, I think "confine_to_keys" is a better name (discuss?)

I think it's better to just make it a list of regexes, rather than strings and regexes.

Also - the implementation of the regex match was a bit odd comparing byte sizes - and Im not convinced Regexp.union is the best way to combine all the regexes, so I've re-written it to match against each regexp individually - at the very least it's a bit more understandable....

I havent had time to test this properly yet, but @quark-zju / @tosmi - does this look like sensible to you?

### CODE REVIEW

**Comment by quark-zju** [comment]
I'm okay with the name change and dropping support for strings.


The current code is inefficient. `lookup` is going to be frequently called. And `@config[:confine_to_keys]` won't change. Therefore

```
+        if @config[:confine_to_keys].is_a?(Array)
+          match = @config[:confine_to_keys].select { |r| key[Regexp.new(r)] }[0]
+          return nil unless match
+        end
```

takes time unnecessarily and produces unnecessary objects for the gc to collect. It should be moved to `initialize`.



And about `Regexp.union`, it should be used for the reason:

* Code is neater. Just remove `is_a?(Array)` test and `Regexp.union` can consume arrays, strings, and regexps.
* Notable faster. See the benchmark:

```ruby
require 'benchmark'

def random_string(len, chars = [*'a'..'z'])
  len.times.map { chars.sample }.join
end

REGEX_N = 10
REGEX_LEN = 40
TEXT_N = 100000
TEXT_LEN = 20

regexs = REGEX_N.times.map { |s| Regexp.new(random_string(REGEX_LEN)) }
regu = Regexp.union(regexs)
texts = TEXT_N.times.map { random_string(TEXT_LEN) }

puts(Benchmark.measure do
  texts.each { |t| t =~ regu }
end) # 0.130000   0.000000   0.130000 (  0.128758)

puts(Benchmark.measure do
  texts.each { |t| regexs.any? { |r| t =~ r } }
end) # 0.280000   0.000000   0.280000 (  0.279627)
```

And about the `bytesize` stuff. I admit it's somehow confusing. But it is there to make sure the matched string is not a substring of key. Otherwise people will have to write `^foo::bar$` instead of `foo::bar`. I think the code can be simplified:

```diff
-key[@key_regexp].to_s.bytesize != key.bytesize
+key[@key_regexp] != key
```

File: lib/hiera/backend/http_backend.rb [hunk 1]
```diff
@@ -38,6 +38,15 @@ def initialize
       end
 
       def lookup(key, scope, order_override, resolution_type)
+
+        # if confine_to_keys is configured, then only proceed if one of the
+        # regexes matches the lookup key
+        #
+        if @config[:confine_to_keys].is_a?(Array)
+          match = @config[:confine_to_keys].select { |r| key[Regexp.new(r)] }[0]
```
Comment (quark-zju) [inline lib/hiera/backend/http_backend.rb]: Note: The "Ruby-ist" way to do `arr.select { ... }[0]` is `arr.find { ... }`. And if we only need a boolean value, use `arr.any? { ... }` is more readable.

**Comment by crayfishx** [comment]
@quark-zju something like that better?

**Comment by quark-zju** [comment]
This version looks good to me :)

**Pull request merged**